// @fridgefm/radio-core/lib/providers/queuestream/queuestream.provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.queuestreamProvider = exports.QUEUESTREAM_TOKEN = void 0;
const stream_1 = require("stream");
const inverter_1 = require("@fridgefm/inverter");
const dev_null_1 = __importDefault(require("dev-null"));
const time_1 = require("../../utils/time");
const events_provider_1 = require("../events/events.provider");
const prebuffer_provider_1 = require("../prebuffer/prebuffer.provider");
const playlist_provider_1 = require("../playlist/playlist.provider");
exports.QUEUESTREAM_TOKEN = (0, inverter_1.createToken)('queuestream');
exports.queuestreamProvider = (0, inverter_1.injectable)({
    provide: exports.QUEUESTREAM_TOKEN,
    scope: 'singleton',
    useFactory: (eventBus, playlist, prebuffer) => {
        let trackStream = new stream_1.Readable();
        const currentStream = new stream_1.Transform({
            transform: (chunk, _, callback) => {
                prebuffer.modify([chunk]);
                callback(undefined, chunk);
            },
        });
        const handleError = (error, event) => {
            eventBus.emit(events_provider_1.PUBLIC_EVENTS.ERROR, {
                name: 'queuestream',
                error,
                event,
            });
            publicQueuestream.next();
        };
        const publicQueuestream = {
            getPrebuffer: () => prebuffer.getStorage(),
            currentPipe: (wrstr, opts) => currentStream.pipe(wrstr, opts),
            next: () => {
                const ct = (0, time_1.captureTime)();
                const nextTrack = playlist.getNext();
                // destroy previous track stream if there was one
                trackStream === null || trackStream === void 0 ? void 0 : trackStream.destroy();
                // populate newly created stream with some handlers
                const [error, newStream] = nextTrack.getSound();
                if (error) {
                    handleError(error, 'get-sound-error');
                    return;
                }
                newStream.once('error', (e) => handleError(e, 'stream-error'));
                newStream.once('end', publicQueuestream.next);
                newStream.pipe(currentStream, { end: false });
                trackStream = newStream;
                eventBus.emit(events_provider_1.PUBLIC_EVENTS.NEXT_TRACK, nextTrack, ct());
            },
        };
        publicQueuestream.currentPipe((0, dev_null_1.default)(), { end: false });
        return publicQueuestream;
    },
    inject: [events_provider_1.EVENT_BUS_TOKEN, playlist_provider_1.PLAYLIST_TOKEN, prebuffer_provider_1.PREBUFFER_TOKEN],
});
