// @fridgefm/inverter/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.modifyToken = exports.createToken = exports.declareModule = exports.injectable = exports.debugContainer = exports.CHILD_DI_FACTORY_TOKEN = exports.declareContainer = void 0;
var container_declaration_1 = require("./module/container-declaration");
Object.defineProperty(exports, "declareContainer", { enumerable: true, get: function () { return container_declaration_1.declareContainer; } });
Object.defineProperty(exports, "CHILD_DI_FACTORY_TOKEN", { enumerable: true, get: function () { return container_declaration_1.CHILD_DI_FACTORY_TOKEN; } });
var debugger_1 = require("./utils/debugger");
Object.defineProperty(exports, "debugContainer", { enumerable: true, get: function () { return debugger_1.debugContainer; } });
var provider_declaration_1 = require("./module/provider-declaration");
Object.defineProperty(exports, "injectable", { enumerable: true, get: function () { return provider_declaration_1.injectable; } });
var module_declaration_1 = require("./module/module-declaration");
Object.defineProperty(exports, "declareModule", { enumerable: true, get: function () { return module_declaration_1.declareModule; } });
var token_1 = require("./base/token");
Object.defineProperty(exports, "createToken", { enumerable: true, get: function () { return token_1.createToken; } });
Object.defineProperty(exports, "modifyToken", { enumerable: true, get: function () { return token_1.modifyToken; } });
