// @fridgefm/radio-core/lib/providers/events/events.provider.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventBusProvider = exports.eventBusFactory = exports.PUBLIC_EVENTS = exports.EVENT_BUS_TOKEN = void 0;
const events_1 = require("events");
const inverter_1 = require("@fridgefm/inverter");
const logger_1 = require("../../utils/logger");
const tokens_1 = require("../tokens");
exports.EVENT_BUS_TOKEN = (0, inverter_1.createToken)('event_bus');
exports.PUBLIC_EVENTS = {
    ERROR: 'error',
    INFO: 'einfo',
    START: 'estart',
    RESTART: 'erestart',
    NEXT_TRACK: 'enexttrack',
};
const eventBusFactory = (config) => {
    const emitter = new events_1.EventEmitter();
    const publicEventBus = {
        on: (eventName, handler) => emitter.on(eventName, handler),
        emit: (eventName, ...args) => emitter.emit(eventName, ...args),
    };
    if (config.verbose) {
        publicEventBus.on(exports.PUBLIC_EVENTS.INFO, (_a) => {
            var { level } = _a, rest = __rest(_a, ["level"]);
            return logger_1.logger.log(level || 'info', Object.assign({}, rest));
        });
        publicEventBus.on(exports.PUBLIC_EVENTS.ERROR, (event) => logger_1.logger.log('error', Object.assign(Object.assign({}, event), { message: event.error.message })));
    }
    publicEventBus.on(exports.PUBLIC_EVENTS.NEXT_TRACK, (tr, timings) => publicEventBus.emit(exports.PUBLIC_EVENTS.INFO, {
        event: exports.PUBLIC_EVENTS.NEXT_TRACK,
        message: tr.fsStats.stringified,
        timings,
    }));
    publicEventBus.on(exports.PUBLIC_EVENTS.START, (_, timings) => publicEventBus.emit(exports.PUBLIC_EVENTS.INFO, {
        event: exports.PUBLIC_EVENTS.START,
        message: 'Station started',
        timings,
    }));
    publicEventBus.on(exports.PUBLIC_EVENTS.RESTART, (_, timings) => publicEventBus.emit(exports.PUBLIC_EVENTS.INFO, {
        event: exports.PUBLIC_EVENTS.RESTART,
        message: 'Station restarted',
        timings,
    }));
    return publicEventBus;
};
exports.eventBusFactory = eventBusFactory;
exports.eventBusProvider = (0, inverter_1.injectable)({
    provide: exports.EVENT_BUS_TOKEN,
    useFactory: exports.eventBusFactory,
    inject: [tokens_1.CONFIG_TOKEN],
});
