// @fridgefm/radio-core/lib/providers/station/station.provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stationProvider = exports.STATION_PUBLIC_TOKEN = void 0;
const inverter_1 = require("@fridgefm/inverter");
const time_1 = require("../../utils/time");
const playlist_provider_1 = require("../playlist/playlist.provider");
const queuestream_provider_1 = require("../queuestream/queuestream.provider");
const events_provider_1 = require("../events/events.provider");
const tokens_1 = require("../tokens");
exports.STATION_PUBLIC_TOKEN = (0, inverter_1.createToken)('station_root');
exports.stationProvider = (0, inverter_1.injectable)({
    provide: exports.STATION_PUBLIC_TOKEN,
    useFactory: (config, queuestream, eventBus, playlist) => {
        const station = {
            start() {
                const ct = (0, time_1.captureTime)();
                if (this.getPlaylist().length) {
                    queuestream.next();
                }
                eventBus.emit(events_provider_1.PUBLIC_EVENTS.START, this.getPlaylist(), ct());
            },
            getPlaylist: () => playlist.getList(),
            addFolder: (folder) => playlist.addFolder(folder),
            next: () => queuestream.next(),
            connectListener(_, res, cb = () => { }) {
                const { currentPipe, getPrebuffer } = queuestream;
                res.writeHead(200, config.responseHeaders);
                res.write(getPrebuffer());
                currentPipe(res);
                cb();
            },
            reorderPlaylist: (cb) => playlist.reorder(cb),
            on: (...args) => eventBus.on(...args),
        };
        return station;
    },
    inject: [tokens_1.CONFIG_TOKEN, queuestream_provider_1.QUEUESTREAM_TOKEN, events_provider_1.EVENT_BUS_TOKEN, playlist_provider_1.PLAYLIST_TOKEN],
});
