// @fridgefm/radio-core/lib/providers/station.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Station = void 0;
const inverter_1 = require("@fridgefm/inverter");
const config_1 = require("./config");
const events_provider_1 = require("./events/events.provider");
const playlist_provider_1 = require("./playlist/playlist.provider");
const queuestream_provider_1 = require("./queuestream/queuestream.provider");
const prebuffer_provider_1 = require("./prebuffer/prebuffer.provider");
const track_provider_1 = require("./track/track.provider");
const station_provider_1 = require("./station/station.provider");
const tokens_1 = require("./tokens");
class Station {
    constructor(extConfig) {
        this.on = (...args) => this._station.on(...args);
        this._station = (0, inverter_1.declareContainer)({
            providers: [
                (0, inverter_1.injectable)({
                    provide: tokens_1.CONFIG_TOKEN,
                    useValue: (0, config_1.createConfig)(extConfig || {}),
                }),
                track_provider_1.trackProvider,
                events_provider_1.eventBusProvider,
                playlist_provider_1.playlistProvider,
                prebuffer_provider_1.prebufferProvider,
                queuestream_provider_1.queuestreamProvider,
                station_provider_1.stationProvider,
            ],
        }).get(station_provider_1.STATION_PUBLIC_TOKEN);
    }
    start() {
        this._station.start();
    }
    addFolder(folder) {
        this._station.addFolder(folder);
    }
    next() {
        this._station.next();
    }
    getPlaylist() {
        return this._station.getPlaylist();
    }
    reorderPlaylist(cb) {
        return this._station.reorderPlaylist(cb);
    }
    connectListener(req, res, cb = () => { }) {
        this._station.connectListener(req, res, cb);
    }
}
exports.Station = Station;
