// @fridgefm/inverter/lib/module/container-declaration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.declareContainer = exports.CHILD_DI_FACTORY_TOKEN = void 0;
const base_container_1 = require("../base/base-container");
const errors_1 = require("../base/errors");
const token_1 = require("../base/token");
const internals_1 = require("../base/internals");
const extractDeclaration = (tokenDeclaration) => {
    if ('token' in tokenDeclaration) {
        return tokenDeclaration;
    }
    return {
        token: tokenDeclaration,
        optional: false,
    };
};
const orderModuleProviders = (modules, events) => {
    const uniqProviders = new Set();
    const traverseModules = (importedModules, parentName) => {
        importedModules === null || importedModules === void 0 ? void 0 : importedModules.forEach((moduleDec) => {
            var _a;
            const { providers, imports = [], name } = moduleDec.__internals;
            traverseModules(imports, name);
            (_a = events === null || events === void 0 ? void 0 : events.moduleRegistered) === null || _a === void 0 ? void 0 : _a.call(events, moduleDec, parentName);
            providers.forEach((providerDec) => {
                var _a;
                if (uniqProviders.has(providerDec)) {
                    // this ensures the correct resolve order - the latest added provider is set for resolve (see tests for additional info)
                    uniqProviders.delete(providerDec);
                }
                uniqProviders.add(providerDec);
                (_a = events === null || events === void 0 ? void 0 : events.providerRegistered) === null || _a === void 0 ? void 0 : _a.call(events, providerDec);
            });
        });
    };
    traverseModules(modules, 'Container');
    return Array.from(uniqProviders.values());
};
/**
 * Token for creating child containers, an injectable should be passed as the root of this container
 */
exports.CHILD_DI_FACTORY_TOKEN = (0, token_1.createToken)('inverter:child-di-factory');
const declareContainer = ({ providers = [], modules = [], parent, events }) => {
    var _a, _b;
    (_a = events === null || events === void 0 ? void 0 : events.containerStart) === null || _a === void 0 ? void 0 : _a.call(events);
    const container = (0, base_container_1.createBaseContainer)(parent);
    const resolvingTokens = new Set(); // for cycle dep check
    const allProviders = [...orderModuleProviders(modules, events), ...providers];
    const traverseProviders = (provider, stack) => {
        var _a;
        (_a = provider.inject) === null || _a === void 0 ? void 0 : _a.forEach((dec) => {
            const { token, optional } = extractDeclaration(dec);
            const provided = allProviders.find((x) => x.provide === token);
            stack.push(token);
            if (typeof provided === 'undefined') {
                if (optional || typeof token.optionalValue !== 'undefined') {
                    return;
                }
                throw new errors_1.ResolverError(stack);
            }
            traverseProviders(provided, stack);
        });
    };
    // @ts-ignore
    container.bindValue(exports.CHILD_DI_FACTORY_TOKEN, (childProvider) => {
        // before we can start creating child di first should check that everything is correctly provided
        // yes, this is a pretty costly operation but we better do it once at start rather than get runtime errors
        traverseProviders(childProvider, [childProvider.provide]);
        return () => (0, exports.declareContainer)({
            parent: container,
            providers: [childProvider],
            events,
        }).get(childProvider.provide);
    });
    allProviders.forEach((injectableDep) => {
        var _a;
        const { inject, provide, useFactory, useValue, scope } = injectableDep;
        if (typeof useFactory !== 'undefined') {
            container.bindFactory(provide, (ctx) => {
                var _a;
                inject === null || inject === void 0 ? void 0 : inject.forEach((tokenDeclaration) => {
                    // check if there is a cycle
                    if (resolvingTokens.has(extractDeclaration(tokenDeclaration).token)) {
                        throw new errors_1.CyclicDepError([provide, ...Array.from(resolvingTokens)]);
                    }
                });
                const resolvedDeps = inject === null || inject === void 0 ? void 0 : inject.map((tokenDeclaration) => {
                    const { token, optional } = extractDeclaration(tokenDeclaration);
                    // mark current resolving dep
                    resolvingTokens.add(token);
                    const resolvedValue = ctx.resolve(token);
                    // delete it after it is resolved
                    resolvingTokens.delete(token);
                    if (resolvedValue === internals_1.NOT_FOUND_SYMBOL) {
                        if (optional) {
                            return undefined;
                        }
                        traverseProviders(injectableDep, [provide]);
                    }
                    return resolvedValue;
                });
                (_a = events === null || events === void 0 ? void 0 : events.providerResolveEnd) === null || _a === void 0 ? void 0 : _a.call(events, injectableDep);
                return useFactory(...(resolvedDeps || []));
            }, { scope });
        }
        else {
            (_a = events === null || events === void 0 ? void 0 : events.providerResolveEnd) === null || _a === void 0 ? void 0 : _a.call(events, injectableDep);
            container.bindValue(provide, useValue);
        }
    });
    (_b = events === null || events === void 0 ? void 0 : events.containerReady) === null || _b === void 0 ? void 0 : _b.call(events);
    return container;
};
exports.declareContainer = declareContainer;
