// @fridgefm/radio-core/lib/providers/prebuffer/prebuffer.provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prebufferProvider = exports.prebufferFactory = exports.PREBUFFER_TOKEN = void 0;
const inverter_1 = require("@fridgefm/inverter");
const tokens_1 = require("../tokens");
exports.PREBUFFER_TOKEN = (0, inverter_1.createToken)('prebuffer');
const prebufferFactory = (config) => {
    const { prebufferLength } = config;
    const storage = [];
    return {
        modify: (chunks) => {
            chunks.forEach((ch) => {
                if (storage.length >= prebufferLength) {
                    storage.shift();
                }
                storage.push(ch);
            });
        },
        getStorage: () => {
            const totalPrebufferLength = (storage[0] || []).length * prebufferLength;
            return Buffer.concat(storage, totalPrebufferLength);
        },
    };
};
exports.prebufferFactory = prebufferFactory;
exports.prebufferProvider = (0, inverter_1.injectable)({
    provide: exports.PREBUFFER_TOKEN,
    scope: 'scoped',
    useFactory: exports.prebufferFactory,
    inject: [tokens_1.CONFIG_TOKEN],
});
