// @fridgefm/inverter/lib/base/base-container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBaseContainer = void 0;
/* eslint-disable @typescript-eslint/no-use-before-define */
const internals_1 = require("./internals");
const errors_1 = require("./errors");
/**
 * Creates a base container where each token and provider have to be manually binded to each other
 */
const createBaseContainer = (parentContainer) => {
    const values = new Map();
    const factories = new Map();
    const multies = new Map();
    const container = {
        bindValue(token, value) {
            if ((0, internals_1.isInternalToken)(token)) {
                return;
            }
            if (token.multi) {
                const prevMulties = multies.get(token.symbol) || [];
                multies.set(token.symbol, [...prevMulties, { value }]);
                return;
            }
            if (factories.has(token.symbol)) {
                factories.delete(token.symbol);
            }
            values.set(token.symbol, value);
        },
        bindFactory(token, factory, options) {
            if ((0, internals_1.isInternalToken)(token)) {
                return;
            }
            if (values.has(token.symbol)) {
                values.delete(token.symbol);
            }
            if (token.multi) {
                const prevMulties = multies.get(token.symbol) || [];
                multies.set(token.symbol, [...prevMulties, { factory, scope: options === null || options === void 0 ? void 0 : options.scope }]);
                return;
            }
            factories.set(token.symbol, { factory, options });
        },
        hasToken(token) {
            var _a;
            return values.has(token.symbol) || factories.has(token.symbol) || ((_a = parentContainer === null || parentContainer === void 0 ? void 0 : parentContainer.hasToken(token)) !== null && _a !== void 0 ? _a : false);
        },
        get(token) {
            const value = resolver(token, container);
            if (value !== internals_1.NOT_FOUND_SYMBOL) {
                return value;
            }
            if (token.optionalValue) {
                return token.optionalValue;
            }
            throw new errors_1.ResolverError([token]);
        },
        resolve(token) {
            const value = resolver(token, container);
            if (value !== internals_1.NOT_FOUND_SYMBOL) {
                return value;
            }
            if (token.optionalValue) {
                return token.optionalValue;
            }
            return internals_1.NOT_FOUND_SYMBOL;
        },
    };
    function resolver(token, origin) {
        var _a;
        if (token.multi) {
            const findInParent = () => {
                const parentResolver = parentContainer === null || parentContainer === void 0 ? void 0 : parentContainer.get(internals_1.INTERNAL_TOKENS.RESOLVER);
                if (parentResolver) {
                    const resolved = parentResolver(token, origin);
                    return Array.isArray(resolved) ? resolved : [];
                }
                return [];
            };
            const multiFactories = multies.get(token.symbol);
            if (typeof multiFactories === 'undefined') {
                return findInParent();
            }
            const resolved = multiFactories.map((multiRecord, index) => {
                if ('value' in multiRecord) {
                    return multiRecord.value;
                }
                const scope = multiRecord.scope || internals_1.DEFAULT_SCOPE;
                if (scope === 'singleton') {
                    // if it is a singleton - cache it
                    const value = multiRecord.factory(container);
                    multiFactories[index] = { value };
                    return value;
                }
                return multiRecord.factory(container);
            });
            return [...resolved, ...findInParent()];
        }
        const arrivedValue = values.get(token.symbol);
        const hasValue = arrivedValue !== undefined || values.has(token.symbol);
        if (hasValue && origin === container) {
            return arrivedValue;
        }
        const factoryContext = factories.get(token.symbol);
        if (factoryContext) {
            const scope = ((_a = factoryContext.options) === null || _a === void 0 ? void 0 : _a.scope) || internals_1.DEFAULT_SCOPE;
            switch (scope) {
                case 'singleton': {
                    if (hasValue) {
                        return arrivedValue;
                    }
                    else if (parentContainer === null || parentContainer === void 0 ? void 0 : parentContainer.hasToken(token)) {
                        break;
                    }
                    else {
                        // Cache the value in the same container where the factory is registered.
                        const value = factoryContext.factory(container);
                        container.bindValue(token, value);
                        return value;
                    }
                }
                case 'scoped': {
                    // Create a value within the origin container and cache it.
                    const value = factoryContext.factory(origin);
                    origin.bindValue(token, value);
                    return value;
                }
                case 'transient': {
                    // Create a value within the origin container and don't cache it.
                    return factoryContext.factory(origin);
                }
            }
        }
        if (hasValue) {
            return arrivedValue;
        }
        const parentResolver = parentContainer === null || parentContainer === void 0 ? void 0 : parentContainer.get(internals_1.INTERNAL_TOKENS.RESOLVER);
        if (parentResolver) {
            return parentResolver(token, origin);
        }
        return internals_1.NOT_FOUND_SYMBOL;
    }
    const bindInternalTokens = () => {
        values.set(internals_1.INTERNAL_TOKENS.CONTAINER.symbol, container);
        values.set(internals_1.INTERNAL_TOKENS.RESOLVER.symbol, resolver);
        values.set(internals_1.INTERNAL_TOKENS.FACTORIES_MAP.symbol, factories);
        if (parentContainer) {
            values.set(internals_1.INTERNAL_TOKENS.PARENT_CONTAINER.symbol, parentContainer);
        }
    };
    bindInternalTokens();
    return container;
};
exports.createBaseContainer = createBaseContainer;
