// @fridgefm/radio-core/lib/providers/playlist/playlist.provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.playlistProvider = exports.PLAYLIST_TOKEN = void 0;
const inverter_1 = require("@fridgefm/inverter");
const fs_1 = require("../../utils/fs");
const time_1 = require("../../utils/time");
const events_provider_1 = require("../events/events.provider");
const track_provider_1 = require("../track/track.provider");
const funcs_1 = require("../../utils/funcs");
const mp3_1 = __importDefault(require("../../utils/mp3"));
exports.PLAYLIST_TOKEN = (0, inverter_1.createToken)('playlist');
exports.playlistProvider = (0, inverter_1.injectable)({
    provide: exports.PLAYLIST_TOKEN,
    scope: 'singleton',
    useFactory: (createTrack, eventBus) => {
        const folders = new Set();
        let currentIndex = -1;
        let tracksMap = new Map();
        let list = [];
        const emitInfo = (a) => {
            eventBus.emit(events_provider_1.PUBLIC_EVENTS.INFO, Object.assign({ name: 'playlist' }, a));
        };
        const revalidate = () => {
            const ct = (0, time_1.captureTime)();
            list = (0, fs_1.createList)(Array.from(folders));
            tracksMap = list
                .filter((path) => {
                const f = (0, funcs_1.extractLast)(path, '/');
                return mp3_1.default.isSupported(f[1]);
            })
                .reduce((acc, path) => {
                // deduplicate if already in map
                if (acc.has(path)) {
                    return acc;
                }
                return acc.set(path, createTrack(path));
            }, new Map());
            const result = publicPlaylist.getList();
            emitInfo({ event: 'revalidate', message: 'Playlist revalidated', timings: ct() });
            return result;
        };
        const publicPlaylist = {
            getList: () => list.map((v, i) => {
                const tra = tracksMap.get(v);
                return Object.assign(Object.assign({}, tra), { isPlaying: currentIndex === i });
            }),
            getNext: () => {
                if (list.length - 1 === currentIndex) {
                    // the playlist drained
                    const ct = (0, time_1.captureTime)();
                    revalidate();
                    currentIndex = 0;
                    eventBus.emit(events_provider_1.PUBLIC_EVENTS.RESTART, publicPlaylist.getList(), ct());
                }
                else {
                    currentIndex += 1;
                }
                const nextPath = list[currentIndex];
                const nextTrack = tracksMap.get(nextPath);
                if (!nextTrack) {
                    emitInfo({ level: 'warn', event: 'no-next-track', message: `No next track found for ${nextPath}` });
                    // try next tracks
                    return publicPlaylist.getNext();
                }
                nextTrack.playCount += 1;
                return Object.assign(Object.assign({}, nextTrack), { isPlaying: true });
            },
            addFolder: (folder) => {
                folders.add(folder);
                return revalidate();
            },
            reorder: (cb) => {
                const ct = (0, time_1.captureTime)();
                const prevList = publicPlaylist.getList();
                const currentlyPlaying = prevList.find((v) => !!v.isPlaying);
                list = cb(prevList).map((b) => b.fsStats.fullPath);
                currentIndex = list.findIndex((v) => v === (currentlyPlaying === null || currentlyPlaying === void 0 ? void 0 : currentlyPlaying.fsStats.fullPath));
                emitInfo({
                    level: 'info',
                    event: 'reorder',
                    message: 'Playlist reordered',
                    timings: ct(),
                });
                return publicPlaylist.getList();
            },
        };
        return publicPlaylist;
    },
    inject: [track_provider_1.TRACK_FACTORY_TOKEN, events_provider_1.EVENT_BUS_TOKEN],
});
