// @fridgefm/inverter/lib/utils/debugger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.debugContainer = void 0;
/* eslint-disable no-console */
const container_declaration_1 = require("../module/container-declaration");
/**
 * This is a basic wrapper to debug the resolving process of your cointainer
 */
const debugContainer = (containerCfg) => {
    const timings = { containerStart: 0, containerReady: 0 };
    const container = (0, container_declaration_1.declareContainer)(Object.assign(Object.assign({}, containerCfg), { events: {
            providerResolveEnd: (provider) => {
                var _a, _b;
                console.log('Token resolved', {
                    desc: provider.provide.symbol.description,
                    type: provider.useFactory ? 'factory' : 'value',
                });
                (_b = (_a = containerCfg.events) === null || _a === void 0 ? void 0 : _a.providerResolveEnd) === null || _b === void 0 ? void 0 : _b.call(_a, provider);
            },
            providerRegistered: (provider) => {
                var _a, _b;
                console.log('Provider registered', {
                    desc: provider.provide.symbol.description,
                    type: provider.useFactory ? 'factory' : 'value',
                });
                (_b = (_a = containerCfg.events) === null || _a === void 0 ? void 0 : _a.providerRegistered) === null || _b === void 0 ? void 0 : _b.call(_a, provider);
            },
            moduleRegistered: (mod, parent) => {
                var _a, _b;
                console.log('Module Registered', { desc: mod.__internals.name, parent });
                (_b = (_a = containerCfg.events) === null || _a === void 0 ? void 0 : _a.moduleRegistered) === null || _b === void 0 ? void 0 : _b.call(_a, mod, parent);
            },
            containerReady: () => {
                var _a, _b;
                timings.containerReady = Date.now();
                const durationMs = timings.containerReady - timings.containerStart;
                console.log(`Container is ready in ${durationMs}ms`);
                (_b = (_a = containerCfg.events) === null || _a === void 0 ? void 0 : _a.containerReady) === null || _b === void 0 ? void 0 : _b.call(_a);
            },
            containerStart: () => {
                timings.containerStart = Date.now();
            },
        } }));
    return { container };
};
exports.debugContainer = debugContainer;
