// @fridgefm/radio-core/lib/utils/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
/* eslint-disable no-param-reassign */
const winston_1 = __importDefault(require("winston"));
const chalk_1 = __importDefault(require("chalk"));
const time_1 = require("./time");
const colors = {
    error: chalk_1.default.redBright,
    info: chalk_1.default.greenBright,
    warn: chalk_1.default.yellowBright,
    default: chalk_1.default.grey,
    time: chalk_1.default.cyan,
};
const fridgeformat = winston_1.default.format((info) => {
    // @ts-ignore
    const chalked = colors[info.level] || colors.default;
    const add = {
        timings: typeof info.timings !== 'undefined' ? `+${info.timings}ms` : '',
        event: `${info.name || ''}:${info.event}::`,
    };
    // @ts-ignore
    info.message = `${colors.time((0, time_1.getHHMMSS)(Date.now()))} ${chalked(add.event)} ${info.message || ''} ${chalked(add.timings)}`;
    return info;
});
exports.logger = winston_1.default.createLogger({
    format: fridgeformat(),
    transports: [
        new winston_1.default.transports.Console({
            format: winston_1.default.format.cli(),
        }),
    ],
});
